cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+	- Columns:
+		- `mutable_column_view&`
+	- Tables:
+		- `mutable_table_view&`
+    - Everything else:
+        - Pass by via raw pointer
+-   Outputs 
+	- Outputs should be *returned*, i.e., no output parameters
+	- Columns:
+		- `std::unique_ptr<column>`
+	- Tables:
+		- `std::unique_ptr<table>`
+    - Scalars:
+        - `std::unique_ptr<scalar>`
+
+
+### Multiple Return Values
+
+Sometimes it is necessary for functions to have multiple outputs. There are a few ways this can be 
+done in C++ (including creating a  `struct`  for the output). One convenient way to do this is 
+using `std::tie`  and  `std::make_pair`. Note that objects passed to `std::make_pair` will invoke 
+either the copy constructor or the move constructor of the object, and it may be preferable to move 
+non-trivially copyable objects (and required for types with deleted copy constructors, like 
+`std::unique_ptr`).
+
+```c++
+std::pair<table, table> return_two_tables(void){
+  cudf::table out0;
+  cudf::table out1;
+  ...
+  // Do stuff with out0, out1
+  
+  // Return a std::pair of the two outputs
+  return std::make_pair(std::move(out0), std::move(out1));
+}
+
+cudf::table out0;
+cudf::table out1;
+std::tie(out0, out1) = cudf::return_two_outputs();
+```
+
+Note:  `std::tuple`  _could_  be used if not for the fact that Cython does not support 
+`std::tuple`. Therefore, libcudf APIs must use  `std::pair`, and are therefore limited to return 
+only two objects of different types. Multiple objects of the same type may be returned via a 
+`std::vector<T>`.
+
+## Iterator-based interfaces
+
+Increasingly, libcudf is moving toward internal (`detail`) APIs with iterator parameters rather 
+than explicit `column`/`table`/`scalar` parameters. As with STL, iterators enable generic 
+algorithms to be applied to arbitrary containers. A good example of this is `cudf::copy_if_else`. 
+This function takes two inputs, and a Boolean mask. It copies the corresponding element from the 
+first or second input depending on whether the mask at that index is true or false. Implementing
+`copy_if_else` for all combinations of `column` and `scalar` parameters is simplified by using
+iterators in the `detail` API.
+
+```c++
+template <typename FilterFn, typename LeftIter, typename RightIter>
+std::unique_ptr<column> copy_if_else(
+  bool nullable,
+  LeftIter lhs_begin,
+  LeftIter lhs_end,
+  RightIter rhs,
+  FilterFn filter,
+  ...);
+```
+`LeftIter` and `RightIter` need only implement the necessary interface for an iterator. libcudf 
+provides a number of iterator types and utilities that are useful with iterator-based APIs from 
+libcudf as well as Thrust algorithms. Most are defined in `include/detail/iterator.cuh`. 
+
+### Pair iterator
+
+The pair iterator is used to access elements of nullable columns as a pair containing an element's 
+value and validity. `cudf::detail::make_pair_iterator` can be used to create a pair iterator from a 
+`column_device_view` or a `cudf::scalar`.
```

Comment:
Should we mention explicitly that `make_pair_iterator` is not available on `mutable_column_device_view`?